### ISSUE TITLE
ITE evaluates all args at the same time

### ISSUE THREAD

**Issue by nkanazawa1989** [sympy/sympy#23295]
In the [API documentation](https://docs.sympy.org/latest/modules/logic.html#sympy.logic.boolalg.ITE) it says
> ITE(A, B, C) evaluates and returns the result of B if A is true else it returns the result of C.

however it likely evaluates `B` even if `A` is false.
```python
import sympy as sym
beta = sym.Symbol("beta")
expr = sym.ITE(sym.Abs(beta) > 0, 1/beta > 1, False)
expr.subs(beta, 0)
```
This yields `TypeError: Invalid comparison of non-real zoo`. We can manually evaluate each `args` but it means we cannot safely lambdify this expression. Seems like these lines are causing the problem.
https://github.com/sympy/sympy/blob/66d7aef9dc1b26055af22e27ba42004c40b95d7c/sympy/core/basic.py#L1076-L1082

**Comment by oscarbenjamin** [comment]
Maybe `ITE` should define `_eval_subs`. I guess there will always be cases though where the first condition does not fully evaluate to True/False but substituting into one of the others might raise.

I'm not sure what the best way to handle this in lambdify is.

**Comment by nkanazawa1989** [comment]
I'm quite new to internal logic of sympy. If I hard-code `False` to `A`, the evaluation of `B` doesn't get called.
```python
import sympy as sym
beta = sym.Symbol("beta")
expr = sym.ITE(False, 1/beta > 1, False)
expr.subs(beta, 0)  # False
```
This indicates, if `ITE` can evaluate `args[0]` or `A` separately, we can solve this problem. Does `_eval_subs` help to implement this?

**Comment by oscarbenjamin** [comment]
Check the intermediate result before calling subs:
```python
In [1]: import sympy as sym
   ...: beta = sym.Symbol("beta")
   ...: expr = sym.ITE(False, 1/beta > 1, False)

In [2]: expr
Out[2]: False
```
When lambdifying the following code is used:
```python
In [3]: expr = sym.ITE(sym.Abs(beta) > 0, 1/beta > 1, False)

In [4]: expr
Out[4]: 
⎧1                 
⎪─ > 1  for │β│ > 0
⎨β                 
⎪                  
⎩False   otherwise 

In [5]: f = lambdify(beta, expr)

In [6]: f(0)
---------------------------------------------------------------------------
ZeroDivisionError                         Traceback (most recent call last)
<ipython-input-6-6bfbdbfff9c4> in <module>
----> 1 f(0)

<lambdifygenerated-1> in _lambdifygenerated(beta)
      1 def _lambdifygenerated(beta):
----> 2     return select([greater(abs(beta), 0),True], [greater(beta**(-1.0), 1),False], default=nan)

ZeroDivisionError: 0.0 cannot be raised to a negative power

In [7]: import inspect

In [8]: print(inspect.getsource(f))
def _lambdifygenerated(beta):
    return select([greater(abs(beta), 0),True], [greater(beta**(-1.0), 1),False], default=nan)
```
The question is (separately from sympy) how would you write the code in the lambdifygenerated function to make this work as intended?

**Comment by nkanazawa1989** [comment]
Thanks @oscarbenjamin I understand how lamdify optimizes the performance (in principle this is the problem of numpy when it's lambdified). Fortunately, this problem has been already investigated in this [stack overflow](https://stackoverflow.com/questions/33248254/numpy-select-lazy-version).

Inspired by the discussion, I confirmed this works.
```python
from numpy import select, greater, nan
def test(beta):
    return select([greater(abs(beta), 0),True], [lambda x: greater(x**(-1.0), 1), False], default=nan)
```

**Comment by oscarbenjamin** [comment]
The solution suggested at stackoverflow involves using a mask array. That seems like a better approach than using select.

**Comment by nkanazawa1989** [comment]
It sounds like piecewise. Indeed it works

```python
import sympy as sym
beta = sym.Symbol("beta")
expr = sym.Piecewise((1/beta > 1, sym.Abs(beta) > 1), (False, True))
expr.subs(beta, 0)  # False
```

Unfortunately it is translated to the same function after lambdify.

**Comment by oscarbenjamin** [comment]
I think `lambdify` needs to be changed to return something that uses mask arrays.

**Comment by nkanazawa1989** [comment]
So we need to update this block?
https://github.com/sympy/sympy/blob/66d7aef9dc1b26055af22e27ba42004c40b95d7c/sympy/printing/numpy.py#L151-L168

**Comment by oscarbenjamin** [comment]
> So we need to update this block?

Probably. Also I think it might be necessary to make changes to other parts of lambdify because as I understand it the printers can only print code expressions whereas what is needed here is a sequence of code statements.

**Comment by ThePauliPrinciple** [comment]
> Probably. Also I think it might be necessary to make changes to other parts of lambdify because as I understand it the printers can only print code expressions whereas what is needed here is a sequence of code statements.

This is correct. Currently `lambdify` prints a string `return ` and appends whatever is the result of the given printer and expression. This would preclude any kind of multiline code from the printer, except for whatever is generated with the `CSE` system.
See also https://github.com/sympy/sympy/issues/22277

**Comment by asmeurer** [comment]
The laqmbdify part of this is basically the same as https://github.com/sympy/sympy/issues/15014
